### PR TITLE
Rich Text Label widget for Help Browser GUI2 port

### LINKED ISSUE #2687: GUI2 rich text label widget for the help browser
We need a rich-text widget using probably an extension to Pango markup (with additional tags) in order to fully port the help browser to GUI2.

The widget has the following requirements:

* Inline images, which can be shown centred between paragraphs of text.
* Floating images, which align to left or right and which text will wrap around. I believe Pango supports this in the form of 'wrap shapes' or something similar.
* Clickable links that trigger a callback, allowing them to link to other pages within the help browser.
* Ability to set tab stops, or some other means of creating tables.

In my opinion the tag names for these should be unchanged, which means `img`, `ref`, `jump` respectively. In particular, don't use `a` for links; it doesn't make sense anyway.

### PR DESCRIPTION
**Pull request by babaissarkar**
Resolves #2687 

Work in progress, full of debug code.

**Work done so far**

-  Update canvas to support setting pango attributes on a part of text. New WML keys in `[text]`
-  Help Markup Parser supports a subset of XML style parsing in addition to existing markup.
-  Tags implemented : `<header>/<h>`, `<bold>/<b>`, `<italic>/<i>`, `<underline>/<u>`, `<format>/<span>`, ``, `<ref>`, `<table>` and related tags. (`<endtable/>`, `<jump/>`, `<br/>`)

**ToDo**
- [x] Correct positioning and wrapping of text and images
- [x] Correct size calculation using `calculate_best_size()`
- [x] Link handling callback
- [x] Correct calculation of Link Position (could be buggy in complex situations)
- [x] Correct wrapping for floating images

### CODE REVIEW

File: src/gui/widgets/rich_label.hpp [hunk 1]
```diff
@@ -0,0 +1,382 @@
...
+}
+
+// ------------ WIDGET -----------{
+
+/**
+ * @ingroup GUIWidgetWML
+ *
+ * A rich_label displays text that can be wrapped but no scrollbars are provided.
+ *
+ * A rich_label has two states because rich_labels are often used as visual indication of the state of the widget it rich_labels.
```
Comment (CelticMinstrel) [inline src/gui/widgets/rich_label.hpp:42]: Looks like overzealous find and replace here.

**Comment by CelticMinstrel** [comment]
Is there a reason why the help markup handling is done in the header?

It seems the "jump" tag is not handled. Do you have an idea of how to handle it, or will we be reworking how we build tables?

File: src/gui/widgets/rich_label.hpp (hunk 1 above)
Comment (babaissarkar) [reply src/gui/widgets/rich_label.hpp:42]: Yes, will fix it.

**Comment by babaissarkar** [comment]
> Is there a reason why the help markup handling is done in the header?
Will be moved to the main file, it's there because I was testing it.
 
> It seems the "jump" tag is not handled. Do you have an idea of how to handle it, or will we be reworking how we build tables?
This PR a work in progress, I'll be adding more commits to it. See my initial comment.
Though I'd like to know how `<jump>` and `<ref>` and `<link>` behave, so that I can correctly implement them. Some specifications will be really good here.

**Comment by babaissarkar** [comment]
> Except all those 100's are actually a value calculated by the code based on the actual width of all the column values. Now, that's obviously a very primitive method of making a table. I don't want to force you to use the same method – the main thing is that there's a way to produce a table at all.

I am thinking about doing something like `<table> row=3 col=3 <data></data><data></data>.....</table>`, and will be done by partioning the available space in a grid like structure. Big texts will get wrapped, but big images might create nuisance. Maybe scale them to fit the area?

**Comment by babaissarkar** [comment]
Working on calculating the correct size for the widget. However, extracting the values from canvas is proving to be really complex, my `get_variable` doesn't work, there are multiple scopes inside `[canvas]` causing the value to disappear. Styled Widget also seems to assume that the widget will always contain text, and calculates the best size that way. The image widget has specialized code that deals with setting the correct size. However, it doesn't work for something that can contain both images and formatted text.
Currently it only works if I hardcode the `default_height/width` in the widget definition file, but clips the image with default size set to zero. (It should calculate the maximum area required to display the widget dynamically, without relying on the default size)
Another point is that we need to actually draw everything on the screen before the maximum required size can be calculated.

**Comment by Pentarctagon** [comment]
The size of images used in the help are generally fixed and known ahead of time though - for example the unit portraits are all scaled to one of a few specific sizes depending on the window size and whether the pixel multiplier is being used.

**Comment by babaissarkar** [comment]
Added a new GUI Test Dialog which can be used to test new widgets and gui2 WML code, and can also be used a template for new dialogs. This is just a blank dialog with an Ok button, and accessible by passing the `--clock` option, which shows the button to open it.
Screenshot from 2024-03-31 16-38-20

File: src/gui/widgets/rich_label.hpp (hunk 1 above)
Comment (babaissarkar) [reply src/gui/widgets/rich_label.hpp]: Fixed.

**Comment by babaissarkar** [comment]
Still many things to do, but here's a small demo :
Screenshot from 2024-04-02 10-30-18

**Comment by babaissarkar** [comment]
> I saw images in the markup but not in the screenshot…

It's commented out.

**Comment by babaissarkar** [comment]
@CelticMinstrel 
Update :
Screenshot from 2024-04-02 20-52-55
Managed to do this. Now the image that you mentioned shows.

**Comment by CelticMinstrel** [comment]
Considering the case of floated images and shaped wrapping again… the idea of the feature was to have it look something like this:

```
............................................
............................................
+---------+ ................................
|         | ................................
|  image  | ................................
|         | ................................
+---------+ ................................
............................................
............................................
```


```
............................................
............................................
................................ +---------+
................................ |         |
................................ |  image  |
................................ |         |
................................ +---------+
............................................
............................................
```

---

But probably one of these would be good enough:

```
            ................................
            ................................
+---------+ ................................
|         | ................................
|  image  | ................................
|         | ................................
+---------+ ................................
            ................................
            ................................
```


```
................................
................................
................................ +---------+
................................ |         |
................................ |  image  |
................................ |         |
................................ +---------+
................................
................................
```

And if even that is too hard, it's not the end of the world if it ends up looking the same as in your screenshot.

**Comment by CelticMinstrel** [comment]
> I've also found a way to calculate text and image sizes, but they need to be drawn first (especially for text).

Hmm, this seems strange.

For text, it should be possible to calculate the size without rendering by adding up the glyph extents, and if pango doesn't provide a function to do that for you (as opposed to manually iterating through the text and querying each glyph individually from FreeType) then I would be rather surprised. (Admittedly, it's a bit more complicated than that in practice, since there isn't a one-to-one correspondence between characters and glyphs – you have combining characters, ligatures, composite characters like the flag emojis, etc.)

For images, I would expect them to be stored on a texture of exactly the right size, so you'd be able to just query the texture size. (Or if they were a sprite, ie a part of a texture, the size would be stored as part of the sprite data. We don't do that right now but it could happen in the future.)

If it doesn't have a performance impact though, then maybe it's fine to do things this way, but it does feel like something is missing.

**Comment by babaissarkar** [comment]
Extremely rudimentary plain text wrapping :
Screenshot from 2024-04-05 12-30-22

Working on improving it.

> I've also found a way to calculate text and image sizes, but they need to be drawn first (especially for text).
Looking once again at my fake draw code, `text_shape` and `image_shape` are not actually rendering or drawing anything for calculating the size, the functions return before that step. `image_shape` also seems to reuse available texture, but not sure about that.

**Comment by babaissarkar** [comment]
`float=yes/no` and `align=left/right` are supported now. However they run into small bugs from time to time.
**More screenshots :**
(`float=yes`)
Screenshot from 2024-04-08 16-34-04

(The version in b8057fc367422b335b258d13702dc5147a66a827 doesn't has `float=yes` added, this screenshot is after adding that)

(no `float` key)

Screenshot from 2024-04-08 16-32-53

I'll start working at the `table` support now, then perhaps `format` and then more bugfixes.

**Comment by CelticMinstrel** [comment]
The float example looks great. If I had one complaint it would be that the text is precisely flush against the image. It might look better if there's a small gap added. The non-float example looks good too. (That list is just build manually with literal bullet characters and tabs, right? No markup involved there.)

**Comment by babaissarkar** [comment]
> The float example looks great. If I had one complaint it would be that the text is precisely flush against the image. It might look better if there's a small gap added. The non-float example looks good too.

I'll add the margin in the final version, don't want to run into another calculation bug ATM.

>  (That list is just build manually with literal bullet characters and tabs, right? No markup involved there.)
Yes, it's manual.

**Comment by babaissarkar** [comment]
Changes in 0042dbe9b2dc7bca554367013a2735981f5821cf :
**1. Basic Table functionality**
Introduces new tags `<table>` (keys : `col`), `<break>` or `<br>` (start new row), `<endtable>`. `<jump>` is repurposed to start a new column.
**2. Parser upgrade**
In order to simplify syntax, the help markup parser has been upgraded to support XML style empty tags with XML attributes. This mode is only active when it encounters an empty tag (tag that ends with '/>'). All existing functionality still works.

### Example
**Old markup :** `<header>text='Movement'</header><table>col=3</table>Hello<jump></jump>Hello2<jump></jump>Hello3<br></br>Hello<jump></jump>Hello2<jump></jump>Hello3<endtable></endtable>Movement in <italic>text='Battle for Wesnoth'</italic> is simple.`
**New markup :** `<header text='Movement'/><table col=3/>Hello<jump/>Hello2<jump/>Hello3<br/>Hello<jump/>Hello2<jump/>Hello3<endtable/>Movement in <italic text='Battle for Wesnoth'/> is simple.`

Screenshot from 2024-04-09 12-42-00

**Comment by CelticMinstrel** [comment]
There is an argument for migrating entirely to XML-style markup, though that's probably out of scope for this PR as it would also entail updating all the existing help topics. The argument is that there are strings that are shown both in the help (requiring help markup) and in places using regular Pango markup. Because of that, these strings that could _theoretically_ use markup must instead _not_ use markup so that they don't break one of the two places they're shown.

I'm not entirely happy with that table markup, but it's definitely a step in the right direction. Any improvements probably require overhauling the entire basis of the help markup, so I'll take what we've got for now.

**Comment by babaissarkar** [comment]
> I'm not entirely happy with that table markup, but it's definitely a step in the right direction. Any improvements probably require overhauling the entire basis of the help markup, so I'll take what we've got for now.

Would be good to have any markup ideas you have written somewhere, so me or someone else guess how much overhauling the parser needs. (Maybe in #2687)

I can probably change the <jump>'s element name to something like <cbr> (column break). It is also quite possible to introduce XML style attributes to non-empty tags. The only nuisance is the nested tag issue you earlier mentioned, that might be harder to implement.

**Comment by CelticMinstrel** [comment]
> Would be good to have any markup ideas you have written somewhere, so me or someone else guess how much overhauling the parser needs.

I don't have any specific ideas, other than that it would be nice if there is a subset of help markup that is also valid Pango markup.

**Comment by babaissarkar** [comment]
Turns out this is possible with minimal effort :
Screenshot from 2024-04-10 13-14-51

The escape by backslash already existed, I simply turned on `text_markup=true`. However the syntax is very weird :
`\<span foreground='red' background='yellow'>Hello\</span>`
You have to quote the `<` but not anything else.

**Comment by CelticMinstrel** [comment]
But does that have the property that it can be put in a rich text label and a normal label with markup and display _roughly_ the same in both?

**Comment by babaissarkar** [comment]
As expected, putting exactly the same markup in a normal label causes the slashes to appear.
Screenshot from 2024-04-10 19-15-27

I'll see if I can get rid of the slashes and make it equivalent on both normal and rich labels.

**Comment by CelticMinstrel** [comment]
I wouldn't make it a priority, but basically the main use-case for this is unit descriptions, which can appear both in the sidebar (which if I recall correctly uses Pango markup) and in the help (which obviously uses help markup). The issue being that there's no markup that's valid in both contexts. (Mainline unit descriptions don't use markup at all, so the problem doesn't arise there.)

**Comment by babaissarkar** [comment]
Is there any UMC content that already uses help markup? Can they add help pages or something like that?

**Comment by CelticMinstrel** [comment]
I think there might actually be some way for add-ons to add topics, but it's not documented (or poorly documented?) so I doubt anyone uses it.

**Comment by babaissarkar** [comment]
> Is there any UMC content that already uses help markup? Can they add help pages or something like that?

Turns out this is issue #1155.

**Comment by babaissarkar** [comment]
### Changes :

- `<span>` tag added. `<format>` now alias to `<span>`. _Supported attributes_ : `foreground/color`, `background/bgcolor`, `size/font_size`
- Help Markup Parser now support XML style attributes in non-empty tags as well.
- Following Pango markup is supported : `<b>`, `<i>` and the `<span>` above.

**Comment by CelticMinstrel** [comment]
> Turns out this is issue #1155.

It is, but I think I remember recently noticing some case where add-ons _can_ actually add custom topics… I don't remember where that was though.

**Comment by babaissarkar** [comment]
@CelticMinstrel
I was trying to build `gui2_help_old` branch to test this, but running into quite a few build errors. Did it used to build?

**Comment by CelticMinstrel** [comment]
That exists mainly for archival purposes in case it still contains something useful. I don't believe it's expected to build. I _think_ `gui2_help` should build though, but I could be wrong there as well. If it doesn't build at the tip, I would expect a few commits back to build.

**Comment by babaissarkar** [comment]
@CelticMinstrel 
> That exists mainly for archival purposes in case it still contains something useful. I don't believe it's expected to build. I _think_ `gui2_help` should build though, but I could be wrong there as well. If it doesn't build at the tip, I would expect a few commits back to build.

Yes, the `gui2_help` branch did build with no problems. Was a bit confused by https://github.com/wesnoth/wesnoth/pull/3653#issuecomment-1403079309 so I tried `gui2_help_old` branch first. When you said that there might be something useful in `gui2_help_old`, were you specifically interested in something that doesn't exists in this branch (`gui2_help`)?

**Comment by CelticMinstrel** [comment]
> When you said that there might be something useful in `gui2_help_old`, were you specifically interested in something that doesn't exists in this branch (`gui2_help`)?

I meant there might be some commits in `gui2_help_old` that are worth cherry-picking / adapting into `gui2_help`.

> I wonder if it would be better to merge this PR into `gui2_help`, complete the help browser, then merge `gui2_help` into master once it's usable.

I'm okay with that strategy, though as you say `gui2_help` needs to be rebased first, plus it would probably be a good idea to double-check nothing important was missed from `gui2_help_old` before doing that. (I think there are commits in `gui2_help_old` that rewrite or even remove the help markup parser; those can be ignored. But it's likely _because_ of those that there could be commits added after them that are still useful)

> Run into trouble with this. `rich_label` expects the signature to be the first one. More adjustments before I can test this.

I think the signature change in ed0615b6a3dc2e6e462c6520e9f34c2f5b761882 is a good one, though I guess we probably don't need the `format_help_text` function that converts help markup to Pango anymore.

**Comment by babaissarkar** [comment]
> I'm okay with that strategy, though as you say `gui2_help` needs to be rebased first, plus it would probably be a good idea to double-check nothing important was missed from `gui2_help_old` before doing that. (I think there are commits in `gui2_help_old` that rewrite or even remove the help markup parser; those can be ignored. But it's likely _because_ of those that there could be commits added after them that are still useful)

Would it be possible for you or somebody else to get started with the rebasing?
 
> I think the signature change in [ed0615b](https://github.com/wesnoth/wesnoth/commit/ed0615b6a3dc2e6e462c6520e9f34c2f5b761882) is a good one, though I guess we probably don't need the `format_help_text` function that converts help markup to Pango anymore.

Agreed.

**Comment by CelticMinstrel** [comment]
> Would it be possible for you or somebody else to get started with the rebasing?

I'll put that on my todo list, though actually doing code is a step up from commenting on issues in terms of motivation required. But then again it _is_ Friday, so maybe I can get it done on the weekend (more likely Sunday than Saturday if so).

**Comment by babaissarkar** [comment]
Managed get the gui2 help browser running with the `rich_label`. That will make looking for bug much easier. Obviously, the markup in the help text will also need to be adjusted.
(I'm not using the new toggle button definition yet)
Screenshot from 2024-04-19 20-49-52

Perhaps I can put this local branch up somewhere to save some work on your part. (Though I don't know where)

**Comment by CelticMinstrel** [comment]
So that branch is 1) gui_help rebased 2) with rich_label merged in?

**Comment by babaissarkar** [comment]
> So that branch is 1) gui_help rebased 2) with rich_label merged in?

I think so, but you'll need to check if I missed anything.

**Comment by CelticMinstrel** [comment]
> (Don't think can be technically called a rebase. No idea how to preserved the commit history. Git help is welcome.)

Hm, maybe it's best if it's used only as a reference then…

**Comment by CelticMinstrel** [comment]
I guess I'd expect floating images to stack vertically and inline images to stack horizontally… and I wouldn't expect an automatic line break after an inline image. Do we actually use inline images in the middle of a line of text though? If it's harder to not have a line break then maybe that's fine.

I think in practice these questions won't matter for the in-game help, as they don't use images extensively enough for those cases to come up… but obviously they still need to be decided.

**Comment by babaissarkar** [comment]
> If "inline" just means it sits as a separate paragraph and text does not flow around it, rather than that it flows with the text as in CSS inline-block, then it's probably fine to add the line break.

No, it doesn't work similarly to `inline-block`, but rather similarly to CSS `display:block` IIRC.
This is what is does. Both images are inline.
Screenshot from 2024-05-15 19-09-24

**Comment by CelticMinstrel** [comment]
It seems like a line break after the image would be expected if it's like CSS `block`.

> Problem is, I have to think the best way to reproduce these

How are those done right now? Could they be done as left-floated images? Or are those in fact inline and you've misunderstood something?

**Comment by babaissarkar** [comment]
> How are those done right now?

Here are the relevant parts :
https://github.com/wesnoth/wesnoth/blob/c57251ef34065aa96890e7fbc08fb9e41054a764/src/help/help_topic_generators.cpp#L650-L656

function `push_tab_pair` :
https://github.com/wesnoth/wesnoth/blob/c57251ef34065aa96890e7fbc08fb9e41054a764/src/help/help_impl.cpp#L1640-L1652

Seems to me it is an `img` followed by a `jump` followed by text.

> Could they be done as left-floated images?
This is what I can currently do :
Screenshot from 2024-05-16 11-29-17

Spacing issues are obvious, so I need something that does exactly as `push_tab_pair`

**Comment by CelticMinstrel** [comment]
> Seems to me it is an `img` followed by a `jump` followed by text.

Seems like a non-floated image?

> Would a "stack" attribute in `img` be useful? It would allow to choose horizontal or vertical stacking when there are multiple images.

Hmm, not sure…

**Comment by babaissarkar** [comment]
> > Seems to me it is an `img` followed by a `jump` followed by text.
> 
> Seems like a non-floated image?

Some experimentation with the current help browser shows that :
1. inline images stack horizontally and insert no linebreaks. Floating images stack vertically.
2. The following code is responsible for the linebreaks :
https://github.com/wesnoth/wesnoth/blob/c57251ef34065aa96890e7fbc08fb9e41054a764/src/help/help_text_area.cpp#L316-L322
My understanding is that if a text block follows an image, it checks if the starting character of the text block is '\n' or not, and if yes, it removes that character and moves the cursor to the correct position.

**Comment by CelticMinstrel** [comment]
> My implementation can do that with \<br/> instead

I fixed this for you (you need a backslash to prevent markdown from parsing the tag).

> Question : logically, should this feature be kept? (the \n one)

I'm not quite sure if I understand what the feature _is_…?

**Comment by babaissarkar** [comment]
> I'm not quite sure if I understand what the feature _is_…?

This :
> If a text block follows an image, it checks if the starting character of the text block is '\n' or not, and if yes, it removes that character and moves the cursor to the correct position.

Should I implement this or do it with `<br/>` instead?

**Comment by CelticMinstrel** [comment]
I meant I didn't understand what that feature is, not that I didn't know what you were referring to. My reading of that description is "it does exactly what you told it to do", so I'm not sure how that counts as a feature to consider implementing or not implementing. Rather, it seems like more of an implementation detail?

**Comment by babaissarkar** [comment]
> Does the latest iteration of the help markup have `<br/>` right now?

Yes.

**Comment by babaissarkar** [comment]
> How does help markup normally treat whitespace? Does it preserve all the whitespace or collapse it like HTML/XML?

Preserves.

**Comment by CelticMinstrel** [comment]
> Except for that one '\n' that I mentioned.

Your description of that sounds like it counts as preserving that one too.

**Comment by babaissarkar** [comment]
> Your description of that sounds like it counts as preserving that one too.

You can call it that. I don't consider it as preserving because it's not being handled by pango, but rather by the help formatting system instead. If pango handled it, it might have appeared somewhere in the middle of image. Anyway, that's just a minor detail, so doesn't matter either way.

**Comment by CelticMinstrel** [comment]
But from the end-user perspective, it ensures that the line break they put in the source appears in the output. So yes, I'd call it an implementation detail of _how_ the line break is preserved.

As for whether to use the same implementation detail in the new system, I think it's fair to leave it up to you – if it makes sense to do the same thing, then you can do it, but if you get the same effect (or close enough) just by leaving the line break in, then that's fine too. Whatever makes the most sense to get the desired output.

There's little point in having `<br/>` if all whitespace is preserved, to be honest. Although, I suppose there could be some edge cases where it's more convenient to use it than literal line breaks, so I don't think it hurts anything to support it.

**Comment by babaissarkar** [comment]
`<br/>` is actually used to start a new row in a table, but as a side effect, it can also introduce manual linebreaks in text.

**Comment by babaissarkar** [comment]
Another demo (partial) :
Screenshot from 2024-05-17 08-22-24

Source markup is in `gui_test_dialog.cfg`.

1. Whitespace between images is significant. They won't be stacked horizontally if there's any text in between.
2. I did show this demo earilier, but this time many of the links are functional.

**Comment by babaissarkar** [comment]
Nested tags not supported by the markup parser.

**Comment by CelticMinstrel** [comment]
I see… it would make sense for it to support nested tags, since it's producing a WML model of the text, but I guess it's not really necessary.

**Comment by babaissarkar** [comment]
What would be a good log domain to put rich_label's logs? I'm talking about the logs that are generated when it reads the help markup and lays out the elements (text, image, table etc.)

**Comment by Pentarctagon** [comment]
Could use the existing `help` log domain, or maybe add something like `help/layout`?

**Comment by soliton-** [comment]
Unless there is an inherit connection to help it should be some domain under `gui/`. `gui/layout` already exist for example or specific for a widget it could be `gui/widgets/rich_label` similar to how there are a couple `gui/dialogs/*` domains.

**Comment by babaissarkar** [comment]
This needs to reviewed for code simplification, if possible.
Also need some help with this :
https://github.com/wesnoth/wesnoth/blob/1ab49a0b856579e21ba6308db6a76e38729bc85c/src/gui/widgets/rich_label.cpp#L202-L263

Most of the bugs now are related to links : in complex situations the link bounding rectangle calculation goes wrong and clicking on the link does nothing (because the clicking rectangle is located somewhere else)

**Comment by babaissarkar** [comment]
> In addition, do you think adding a `border` property might help? Don't know if the `box` key in `` is similar.

A look in Wesnoth 1.6 showed that `box` used to draw a black border around an image. Default value was `yes`.
Screenshot from 2024-06-10 11-48-59

**Pull request merged**